src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableList_1.cs

```diff
@@ -2205,18 +2205,28 @@ internal Node RemoveAll(Predicate<T> match)
                 Contract.Ensures(Contract.Result<Node>() != null);
 
                 var result = this;
-                int index = 0;
-                foreach (var item in this)
+                var enumerator = new Enumerator(result);
+                try
                 {
-                    if (match(item))
+                    var startIndex = 0;
+                    while(enumerator.MoveNext())
                     {
-                        result = result.RemoveAt(index);
-                    }
-                    else
-                    {
-                        index++;
+                        if (match(enumerator.Current))
+                        {
+                            result = result.RemoveAt(startIndex);
+                            enumerator.Dispose();
```

Comment:
Strictly speaking, if after disposal, an exception is thrown (say, from the `Enumerator` constructor) then there's a danger that `enumerator` will be disposed *twice*, which might be bad if we're pooling enumerator objects internally. So perhaps it would be best if after your `Dispose()` call, you explicitly clear the struct with `enumerator = default(Enumerator)` before calling the constructor.